### PR TITLE
Retrieve metada in case it is missing

### PR DESCRIPTION
**Pull request by matei-oltean**

### CODE REVIEW

File: fileutils/extensions.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		return nil, 0, err
+	}
+	if ben.Dict == nil {
+		return nil, 0, errors.New("nil dictionary in payload")
+	}
+	msgType, ok := ben.Dict["msg_type"]
+	if !ok {
+		return nil, 0, errors.New("payload missing \"msg_type\" entry")
+	}
+	if messaging.MessageType(msgType.Int) == messaging.EReject {
```
Comment (matei-oltean) [inline fileutils/extensions.go]: `messaging` should not have to be imported in this package

File: messaging/messages.go [hunk 2]
```diff
@@ -20,6 +20,14 @@ const (
 	MRequest
 	MPiece
 	MCancel
+	MExtended MessageType = 20
+)
+
+// Extension messages
+const (
```
Comment (matei-oltean) [inline messaging/messages.go]: Move this to `fileutils`

File: peer/peer.go [hunk 3]
```diff
@@ -114,15 +153,37 @@ func parsePiece(payload []byte) (*chunk, error) {
...
+		return nil, fmt.Errorf("expected message of length at least 1 got %d instead", len(payload))
+	}
+	parsed, index, err := fileutils.ParseExtensionsMetadata(payload[1:])
+	if err != nil || parsed == nil {
+		return nil, err
+	}
+	return &chunk{
+		chunkType: cInfo,
+		index:     index,
+		begin:     index * chunkSize,
```
Comment (matei-oltean) [inline peer/peer.go]: All metainfo messages should be 16 KiB long (minus the last); create another constant to make it clearer?

File: peer/peer.go [hunk 4]
```diff
@@ -213,31 +288,60 @@ func Download(handshake []byte, address string, pieces chan *Piece, results chan
...
+			if !bytes.Equal(hash[:], h[:]) {
+				needsInfo = true
+				continue
+			}
+			inf, err := fileutils.ParseInfo(res, hash)
+			if err != nil {
+				log.Printf("Disconnecting from peer at %s: %s", address, err)
+				return
+			}
+			info <- inf
```
Comment (matei-oltean) [inline peer/peer.go]: only set `needsInfo` to false here.

File: fileutils/extensions.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	if err != nil {
+		return nil, 0, err
+	}
+	if ben.Dict == nil {
+		return nil, 0, errors.New("extension message has no dictionary")
+	}
+	dict := ben.Dict
+	m, ok := dict["m"]
+	if !ok || m.Dict == nil {
+		return nil, 0, errors.New("extension message has no \"m\" key")
```
Comment (mookid) [inline fileutils/extensions.go]: an opportunity to factor?

**Pull request merged**